### ISSUE TITLE
Crash with "`@:' is not allowed as an instance variable name"

### ISSUE THREAD

**Issue by tkrotoff** [gjtorikian/html-proofer#173]
Using v2.1.0 with a Jekyll website and the Rake snippet from the documentation. (gems installed inside vendor/bundle).

```
Running ["ImageCheck", "LinkCheck", "ScriptCheck"] checks on ./_site on *.html... 


rake aborted!
`@:' is not allowed as an instance variable name
/vendor/bundle/ruby/2.0.0/gems/html-proofer-2.1.0/lib/html/proofer/checkable.rb:13:in `instance_variable_set'
/vendor/bundle/ruby/2.0.0/gems/html-proofer-2.1.0/lib/html/proofer/checkable.rb:13:in `block in initialize'
/vendor/bundle/ruby/2.0.0/gems/html-proofer-2.1.0/lib/html/proofer/checkable.rb:12:in `each_pair'
/vendor/bundle/ruby/2.0.0/gems/html-proofer-2.1.0/lib/html/proofer/checkable.rb:12:in `initialize'
/vendor/bundle/ruby/2.0.0/gems/html-proofer-2.1.0/lib/html/proofer/checks/links.rb:31:in `new'
/vendor/bundle/ruby/2.0.0/gems/html-proofer-2.1.0/lib/html/proofer/checks/links.rb:31:in `block in run'
/vendor/bundle/ruby/2.0.0/gems/nokogiri-1.6.6.2/lib/nokogiri/xml/node_set.rb:187:in `block in each'
/vendor/bundle/ruby/2.0.0/gems/nokogiri-1.6.6.2/lib/nokogiri/xml/node_set.rb:186:in `upto'
/vendor/bundle/ruby/2.0.0/gems/nokogiri-1.6.6.2/lib/nokogiri/xml/node_set.rb:186:in `each'
/vendor/bundle/ruby/2.0.0/gems/html-proofer-2.1.0/lib/html/proofer/checks/links.rb:30:in `run'
/vendor/bundle/ruby/2.0.0/gems/html-proofer-2.1.0/lib/html/proofer.rb:121:in `block (2 levels) in check_files_for_internal_woes'
/vendor/bundle/ruby/2.0.0/gems/html-proofer-2.1.0/lib/html/proofer.rb:118:in `each'
/vendor/bundle/ruby/2.0.0/gems/html-proofer-2.1.0/lib/html/proofer.rb:118:in `block in check_files_for_internal_woes'
/vendor/bundle/ruby/2.0.0/gems/parallel-1.3.4/lib/parallel.rb:430:in `call'
/vendor/bundle/ruby/2.0.0/gems/parallel-1.3.4/lib/parallel.rb:430:in `call_with_index'
/vendor/bundle/ruby/2.0.0/gems/parallel-1.3.4/lib/parallel.rb:336:in `process_incoming_jobs'
/vendor/bundle/ruby/2.0.0/gems/parallel-1.3.4/lib/parallel.rb:318:in `block in worker'
/vendor/bundle/ruby/2.0.0/gems/parallel-1.3.4/lib/parallel.rb:311:in `fork'
/vendor/bundle/ruby/2.0.0/gems/parallel-1.3.4/lib/parallel.rb:311:in `worker'
/vendor/bundle/ruby/2.0.0/gems/parallel-1.3.4/lib/parallel.rb:299:in `block in create_workers'
/vendor/bundle/ruby/2.0.0/gems/parallel-1.3.4/lib/parallel.rb:298:in `each'
/vendor/bundle/ruby/2.0.0/gems/parallel-1.3.4/lib/parallel.rb:298:in `create_workers'
/vendor/bundle/ruby/2.0.0/gems/parallel-1.3.4/lib/parallel.rb:255:in `work_in_processes'
/vendor/bundle/ruby/2.0.0/gems/parallel-1.3.4/lib/parallel.rb:191:in `map'
/vendor/bundle/ruby/2.0.0/gems/html-proofer-2.1.0/lib/html/proofer.rb:114:in `check_files_for_internal_woes'
/vendor/bundle/ruby/2.0.0/gems/html-proofer-2.1.0/lib/html/proofer.rb:100:in `check_directory_of_files'
/vendor/bundle/ruby/2.0.0/gems/html-proofer-2.1.0/lib/html/proofer.rb:75:in `run'
/Rakefile:5:in `block in <top (required)>'
Tasks: TOP => test
(See full trace by running task with --trace)
```

**Comment by gjtorikian** [comment]
Can you show the output of the site in question? It'd help to see the generated HTML so I know where it went wrong.

**Comment by tkrotoff** [comment]
Here one of the page where I can reproduce the crash: http://pastebin.com/qqwBanZH
W3C validator tells me my HTML is not valid with 52 errors.

**Comment by tkrotoff** [comment]
Ok I got the error. Line 147:
```
href="mailto:?subject=La méthode "De Gasquet" pour se préparer
```
Double quotes present inside href=""

**Comment by tkrotoff** [comment]
btw HTML-Proofer returns ` * mailto:?subject=Hello contain an invalid email address`
Probably a good idea to not check the email format when you have a href that starts with `mailto:?subject=` or `mailto:?body=`

**Comment by penibelst** [comment]
@tkrotoff Invalid input may produce invalid reports.

@gjtorikian The two errors above are produced by:

```ruby
add_issue("#{link.href} contains no email address", l.line) if link.path.empty?
add_issue("#{link.href} contain an invalid email address", l.line) unless link.path.include?('@')
```

I propose to exit the method when the path is empty.

**Comment by tkrotoff** [comment]
@penibelst a crash resulting from an invalid input is a bit harsh for the users and difficult to debug: I've spent a lot of time figuring out which page and code snippet was provoking this.

**Comment by mathie** [comment]
I'm seeing a similar error when I run htmlproof on a Jekyll web site:

    > htmlproof _site
    Running ["ScriptCheck", "LinkCheck", "ImageCheck"] checks on _site on *.html... 

    htmlproof 2.1.0 | Error:  `@xmlns:cc' is not allowed as an instance variable name


Given the error, I suspect the line causing the issue is:

    <a xmlns:cc="http://creativecommons.org/ns#" href="https://woss.name/" property="cc:attributionName" rel="cc:attributionURL">Graeme Mathieson</a>

Am I supplying invalid HTML?

**Comment by penibelst** [comment]
> Am I supplying invalid HTML?

I think so. You can validate your markup in the official [validator](https://html5.validator.nu/). Where is the cc-stuff come from?

**Comment by mathie** [comment]
It’s the HTML code that’s generated by the [Creative Commons license chooser](http://creativecommons.org/choose/), which spits out the following:

```html
<a rel="license" href="http://creativecommons.org/licenses/by/4.0/“>
  <img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by/4.0/88x31.png" />
</a><br />
<span xmlns:dct="http://purl.org/dc/terms/" href="http://purl.org/dc/dcmitype/Text" property="dct:title" rel="dct:type”>
  Notes from a Messy Desk
</span>
by
<a xmlns:cc="http://creativecommons.org/ns#" href="http://woss.name/" property="cc:attributionName" rel="cc:attributionURL”>
  Graeme Mathieson
</a>
is licensed under a
<a rel="license" href="http://creativecommons.org/licenses/by/4.0/“>
  Creative Commons Attribution 4.0 International License
</a>.
```

The validator does complain about it, right enough:

    Warning: Attribute with the local name xmlns:cc is not serializable as XML 1.0.

but it seems strange that Creative Commons would be encouraging people to attach invalid HTML to their pages!

**Comment by penibelst** [comment]
I would remove all the crazy attributes. They will not gain the value of your website.

**Comment by gjtorikian** [comment]
I agree that Proofer should provide a more useful message than `@:' is not allowed as an instance variable name`--at least, the file where it's occurring, or the attribute that's messed up. I will investigate adding that. However...

> Probably a good idea not to check the email address when you have a href that starts with mailto:?subject= or mailto:?body=

As far as I can tell from [MDN](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Email_links), `mailto` should be followed immediately by the email address. Any parameters such as `subject` or `body` can come afterward. I don't know where the `mailto` would go if an email address is not present, so I would prefer not to skip them. 

> but it seems strange that Creative Commons would be encouraging people to attach invalid HTML to their pages!

For shaaaaaame!

**Comment by penibelst** [comment]
@gjtorikian Have you found why it crashes? Nokogiri can create the document object:

```ruby
Nokogiri::HTML('<a xmlns:cc="http://creativecommons.org/ns#">Graeme</a>')
=> #<Nokogiri::HTML::Document:0x43263e8 name="document" children=[#<Nokogiri::XML::DTD:0x4315eb2 name="html">, #<Nokogiri::XML::Element:0x4314d78 name="html" children=[#<Nokogiri::XML::Element:0x4314ac6 name="body" children=[#<Nokogiri::XML::Element:0x4314896 name="a" attributes=[#<Nokogiri::XML::Attr:0x43147b0 name="xmlns:cc" value="http://creativecommons.org/ns#">] children=[#<Nokogiri::XML::Text:0x43141ac "Graeme">]>]>]>]>
```

**Comment by gjtorikian** [comment]
Yeah, it's because of Ruby not liking `:xmlns:cc` as an attribute name. From this line of code: https://github.com/gjtorikian/html-proofer/blob/dfcc1a43c3d6b1ab483ae85a03c232b55e79db95/lib/html/proofer/checkable.rb#L12-L14

``` ruby
        obj.attributes.each_pair do |attribute, value|
          instance_variable_set("@#{attribute.tr('-', '_')}".to_sym, value.value)
        end
```

While doing 2.0, I thought it would be useful for people writing custom tests to have access to every variable. Maybe instead of https://github.com/gjtorikian/html-proofer/pull/179 the right approach is to just return these attributes as strings, not symbols.

**Comment by dsernst** [comment]
Still got this crash without any more helpful error reporting:

```
htmlproof 2.4.2 | Error:  `@yourself.' is not allowed as an instance variable name
```

I was able to track it down by re-running htmlproof with the `--verbose` flag to figure out how far it got before it crashed, and then manually inspecting that page I found `mailto:` links with unescaped `"` breaking the link. 

So this particular instance is fixed. But it would have been much easier if htmlproof could've pointed me in the right direction sooner.
